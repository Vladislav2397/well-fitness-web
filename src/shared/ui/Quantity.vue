<template lang="pug">

.quantity(
    :class="`quantity--${state.class}`"
)
    | {{ state.label }}
    ._list
        ._dot(
            v-for="i in 3"
            :key="i"
            :class="{ 'quantity__dot--fill' : i <= count }"
        )

</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'

@Component
export default class Quantity extends Vue {
    @Prop() readonly count!: 0 | 1 | 2 | 3

    get state(): { label: string, class: string } {
        switch (this.count) {
        case 0:
            return {
                label: 'Нет в наличии',
                class: 'error',
            }
        case 1:
            return {
                label: 'Осталось мало',
                class: 'warning',
            }
        default:
            return {
                label: 'В наличии',
                class: 'success'
            }
        }
    }
}

</script>

<style lang="scss">
.quantity {
    color: var(--color);

    &--error {
        --color: #f53b49;
    }

    &--warning {
        --color: #f99808; // FIXME: Make a var
    }

    &--success {
        --color: #2fc509;
    }

    &__list {
        display: inline-block;
        margin-left: toRem(6);
    }

    &__dot {
        width: toRem(6);
        height: toRem(6);
        border: 1px solid;
        border-radius: 50%;
        display: inline-block;

        &--fill {
            background-color: var(--color);
        }

        & + & {
            margin-left: toRem(2);
        }
    }
}
</style>
