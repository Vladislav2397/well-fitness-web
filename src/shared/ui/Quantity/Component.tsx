import { Component, Prop, Vue } from 'vue-property-decorator'
import { CreateElement, RenderContext } from 'vue'

export type QuantityProps = {
    count: QuantityType
}

type QuantityType = 0 | 1 | 2 | 3

const getState = (count: QuantityType) => {
    switch (count) {
        case 0:
            return {
                label: 'Нет в наличии',
                class: 'error',
            }
        case 1:
            return {
                label: 'Осталось мало',
                class: 'warning',
            }
        default:
            return {
                label: 'В наличии',
                class: 'success',
            }
    }
}

@Component({
    // @ts-ignore
    functional: true,
})
export default class Quantity extends Vue {
    @Prop() readonly count!: QuantityType

    render(
        h: CreateElement,
        { props: { count } }: RenderContext<QuantityProps>
    ) {
        const state = getState(count)

        return (
            <div class={['quantity', `quantity--${state.class}`]}>
                {state.label}
                <div class="quantity__list">
                    {[1, 2, 3].map(i => (
                        <div
                            class={[
                                'quantity__dot',
                                { 'quantity__dot--fill': i <= count },
                            ]}
                            key={i}></div>
                    ))}
                </div>
            </div>
        )
    }
}
