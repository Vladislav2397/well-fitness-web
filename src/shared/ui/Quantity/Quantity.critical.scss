.quantity {
    color: var(--color);

    &--error {
        --color: #f53b49;
    }

    &--warning {
        --color: #f99808; // FIXME: Make a var
    }

    &--success {
        --color: #2fc509;
    }

    &__list {
        display: inline-block;
        margin-left: toRem(6);
    }

    &__dot {
        width: toRem(6);
        height: toRem(6);
        border: 1px solid;
        border-radius: 50%;
        display: inline-block;

        &--fill {
            background-color: var(--color);
        }

        & + & {
            margin-left: toRem(2);
        }
    }
}
