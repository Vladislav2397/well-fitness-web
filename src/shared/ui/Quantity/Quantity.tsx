import { RenderContext, VNode } from 'vue'
import * as tsx from 'vue-tsx-support'
import { RecordPropsDefinition } from 'vue/types/options'

export type QuantityProps = {
    count: QuantityType
}

type QuantityType = 0 | 1 | 2 | 3

const getState = (count: QuantityType) => {
    switch (count) {
        case 0:
            return {
                label: 'Нет в наличии',
                class: 'error',
            }
        case 1:
            return {
                label: 'Осталось мало',
                class: 'warning',
            }
        default:
            return {
                label: 'В наличии',
                class: 'success',
            }
    }
}

const QuantityVue = tsx.component<
    QuantityProps,
    RecordPropsDefinition<QuantityProps>
>({
    functional: true,
    name: 'Quantity',
    render(h, { props: { count } }) {
        const state = getState(count)

        return (
            <div class={['quantity', `quantity--${state.class}`]}>
                {state.label}
                <div class="quantity__list">
                    {[1, 2, 3].map(i => (
                        <div
                            class={[
                                'quantity__dot',
                                { 'quantity__dot--fill': i <= count },
                            ]}
                            key={i}
                        />
                    ))}
                </div>
            </div>
        )
    },
})

const Quantity = ({
    props: { count },
}: RenderContext<QuantityProps>): VNode | VNode[] => {
    const state = getState(count)

    return (
        <div class={['quantity', `quantity--${state.class}`]}>
            {state.label}
            <div class="quantity__list">
                {[1, 2, 3].map(i => (
                    <div
                        class={[
                            'quantity__dot',
                            { 'quantity__dot--fill': i <= count },
                        ]}
                        key={i}
                    />
                ))}
            </div>
        </div>
    )
}

export default tsx.ofType<QuantityProps, unknown>().convert(Quantity as any)
