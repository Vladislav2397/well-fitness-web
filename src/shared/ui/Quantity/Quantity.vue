<template lang="pug">

.b-quantity(
    :class="`quantity--${state.class}`"
)
    | {{ state.label }}
    .__list
        .__dot(
            v-for="i in 3"
            :key="i"
            :class="{ 'quantity__dot--fill' : i <= count }"
        )

</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'

export type QuantityProps = Pick<Quantity, 'count'>

@Component
export default class Quantity extends Vue {
    @Prop() readonly count!: 0 | 1 | 2 | 3

    get state(): { label: string; class: string } {
        switch (this.count) {
            case 0:
                return {
                    label: 'Нет в наличии',
                    class: 'error',
                }
            case 1:
                return {
                    label: 'Осталось мало',
                    class: 'warning',
                }
            default:
                return {
                    label: 'В наличии',
                    class: 'success',
                }
        }
    }
}
</script>

<!--<style lang="scss" src="./Quantity.critical.scss"/>-->
<!--<style lang="scss" src="./Quantity.main.scss"/>-->
