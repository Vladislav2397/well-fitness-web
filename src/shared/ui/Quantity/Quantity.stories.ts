import { Story } from '@storybook/vue'
import Quantity, { type QuantityProps } from './Quantity'

export default {
    title: 'shared/Quantity',
    component: Quantity,
}

const Template: Story<QuantityProps> = (args, { argTypes }) => ({
    props: Object.keys(argTypes),
    components: { Quantity },
    methods: {},
    template: '<Quantity v-bind="$props"/>',
})

export const Default = Template.bind({})
Default.args = {
    count: 3,
}
