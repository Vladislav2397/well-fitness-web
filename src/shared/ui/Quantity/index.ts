import Quantity, { type QuantityProps } from './Quantity'

export { Quantity, QuantityProps }
