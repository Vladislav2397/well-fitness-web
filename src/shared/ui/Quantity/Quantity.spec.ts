import { shallowMount } from '@vue/test-utils'
import { Quantity } from './Quantity'

describe('Quantity component.spec', () => {
    test.each([0, 1, 2, 3])('Snapshots for count=%s', count => {
        const wrapper = shallowMount(Quantity, {
            propsData: {
                count,
            },
        })
        expect(wrapper).toMatchSnapshot()
    })
})
