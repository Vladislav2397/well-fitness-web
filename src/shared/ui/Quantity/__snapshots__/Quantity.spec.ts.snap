// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quantity component.spec Snapshots for count 0 1`] = `
<div class="quantity quantity--error">Нет в наличии<div class="quantity__list">
    <div class="quantity__dot"></div>
    <div class="quantity__dot"></div>
    <div class="quantity__dot"></div>
  </div>
</div>
`;

exports[`Quantity component.spec Snapshots for count 1 1`] = `
<div class="quantity quantity--warning">Осталось мало<div class="quantity__list">
    <div class="quantity__dot quantity__dot--fill"></div>
    <div class="quantity__dot"></div>
    <div class="quantity__dot"></div>
  </div>
</div>
`;

exports[`Quantity component.spec Snapshots for count 2 1`] = `
<div class="quantity quantity--success">В наличии<div class="quantity__list">
    <div class="quantity__dot quantity__dot--fill"></div>
    <div class="quantity__dot quantity__dot--fill"></div>
    <div class="quantity__dot"></div>
  </div>
</div>
`;

exports[`Quantity component.spec Snapshots for count 3 1`] = `
<div class="quantity quantity--success">В наличии<div class="quantity__list">
    <div class="quantity__dot quantity__dot--fill"></div>
    <div class="quantity__dot quantity__dot--fill"></div>
    <div class="quantity__dot quantity__dot--fill"></div>
  </div>
</div>
`;

exports[`Quantity component.spec Snapshots for count=0 1`] = `
<div class="quantity quantity--error">Нет в наличии<div class="quantity__list">
    <div class="quantity__dot"></div>
    <div class="quantity__dot"></div>
    <div class="quantity__dot"></div>
  </div>
</div>
`;

exports[`Quantity component.spec Snapshots for count=1 1`] = `
<div class="quantity quantity--warning">Осталось мало<div class="quantity__list">
    <div class="quantity__dot quantity__dot--fill"></div>
    <div class="quantity__dot"></div>
    <div class="quantity__dot"></div>
  </div>
</div>
`;

exports[`Quantity component.spec Snapshots for count=2 1`] = `
<div class="quantity quantity--success">В наличии<div class="quantity__list">
    <div class="quantity__dot quantity__dot--fill"></div>
    <div class="quantity__dot quantity__dot--fill"></div>
    <div class="quantity__dot"></div>
  </div>
</div>
`;

exports[`Quantity component.spec Snapshots for count=3 1`] = `
<div class="quantity quantity--success">В наличии<div class="quantity__list">
    <div class="quantity__dot quantity__dot--fill"></div>
    <div class="quantity__dot quantity__dot--fill"></div>
    <div class="quantity__dot quantity__dot--fill"></div>
  </div>
</div>
`;
